### PR TITLE
esp32/machine_timer: Restrict timer numbers for ESP32C6 to 0 and 1.

### LINKED ISSUE #16438: ESP32C6 Timers: Only Timer(0) and Timer(2) are functional for ESP32-C6 v1.24
### Port, board and/or hardware

esp32-c6, generic board

### MicroPython version

Micropython 1.24 custom compile only adding graphics library that does not use timer

### Reproduction

not (likely) a build issue (I did not test on stock release firmware)

### Expected behaviour

examples from Micropython documentation on Timer class:

WORKS CORRECTLY:
from machine import Timer
tim0 = Timer(0)
tim0.init(period=5000, mode=Timer.ONE_SHOT, callback=lambda t:print(0))

FAILS:
from machine import Timer
tim1 = Timer(1)
tim1.init(period=2000, mode=Timer.ONE_SHOT, callback=lambda t:print(1))

WORKS CORRECTLY:
from machine import Timer
tim2 = Timer(2)
tim2.init(period=5000, mode=Timer.ONE_SHOT, callback=lambda t:print(2))

FAILS:
from machine import Timer
tim3 = Timer(3)
tim3.init(period=2000, mode=Timer.ONE_SHOT, callback=lambda t:print(3))

### Observed behaviour

code for Timer0 returns a '0'
code for Timer1 returns nothing
code for Timer2 returns a '2'
code for Timer3 returns nothing

### Additional Information

The fix is to make the timer code for the ESP32-C6 behave like the ESP32-C3:

it appears the fix should be in the esp32 port file `machine_timer.c` line 79 which should be changed from:
`#if CONFIG_IDF_TARGET_ESP32C3`
to
`#if CONFIG_IDF_TARGET_ESP32C3 or CONFIG_IDF_TARGET_ESP32C6` (or something like that because I do not know the syntax)

### Code of Conduct

Yes, I agree

### PR DESCRIPTION
**Pull request by robert-hh**
### Summary

The ESP32C6 has only one timer in each of the two groups, like the ESP32C3.
Add also a check for valid timer numbers.

Addresses the issue #16438 raised by @mattklapman.

### Testing

Tested with an ESP32C6 generic board and a ESP32 Generic board.

### CODE REVIEW

**Review by projectgus — COMMENTED** [review]
Thanks for picking this up, @robert-hh. If you don't have time to apply and test these suggestions then let me know, I think it'd be fine to merge this fix now and then move over to SOC macros later.

**Comment by mattklapman** [comment]
Don't forget line 69 identical fix in addition to line 79

**Comment by robert-hh** [comment]
That was already mentioned in my comment.

File: ports/esp32/machine_timer.c [hunk 1]
```diff
@@ -108,7 +108,16 @@ static mp_obj_t machine_timer_make_new(const mp_obj_type_t *type, size_t n_args,
...
 
     // Create the new timer.
-    machine_timer_obj_t *self = machine_timer_create(mp_obj_get_int(args[0]));
+    uint32_t timer_number = mp_obj_get_int(args[0]);
+    #if CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32C6
+    uint32_t timer_count = 2;
+    #else
+    uint32_t timer_count = 4;
+    #endif
+    if (timer_number >= timer_count) {
```
[the remark this answers is not in the document]
Comment (robert-hh) [reply ports/esp32/machine_timer.c:117]: Thank Angus. These are very useful Macros and I will change the PR to use them.

**Comment by robert-hh** [comment]
Updated. Tested with ESP32, ESP32S3, ESP32C3, ESP32C6.

**Review by projectgus — APPROVED** [review]
Thanks @robert-hh!

File: ports/esp32/machine_timer.c [hunk 2]
```diff
@@ -108,7 +108,11 @@ static mp_obj_t machine_timer_make_new(const mp_obj_type_t *type, size_t n_args,
     mp_arg_check_num(n_args, n_kw, 1, MP_OBJ_FUN_ARGS_MAX, true);
 
     // Create the new timer.
-    machine_timer_obj_t *self = machine_timer_create(mp_obj_get_int(args[0]));
+    uint32_t timer_number = mp_obj_get_int(args[0]);
+    if (timer_number >= SOC_TIMER_GROUP_TOTAL_TIMERS) {
+        mp_raise_ValueError(MP_ERROR_TEXT("Invalid Timer number"));
```
Comment (dpgeorge) [inline ports/esp32/machine_timer.c:113]: Please start error messages with lowercase (this is for consistency, and also improves the error message compression).

Comment (robert-hh) [reply ports/esp32/machine_timer.c]: Done & updated.

**Comment by dpgeorge** [comment]
@robert-hh you have a stray change to the `lib/pico-sdk` submodule here.

File: ports/esp32/machine_timer.c (hunk 2 above)
Comment (dpgeorge) [reply ports/esp32/machine_timer.c]: Thanks!

**Review by dpgeorge — APPROVED** [review]

**Comment by dpgeorge** [comment]
Thanks for updating, it's all good now.

**Comment by graeme44** [comment]
@robert-hh @projectgus, I'm not so sure about this. I'm testing this now and the ESP32C3 loses access to two timers.  

> [The two ESP32-C3 timer groups, with two timers in each, provide the total of four individual timers for use.](https://docs.espressif.com/projects/esp-idf/en/v4.3/esp32c3/api-reference/peripherals/timer.html)

**Comment by graeme44** [comment]
> @robert-hh @projectgus, I'm not so sure about this. I'm testing this now and the ESP32C3 loses access to two timers.
> 
> > [The two ESP32-C3 timer groups, with two timers in each, provide the total of four individual timers for use.](https://docs.espressif.com/projects/esp-idf/en/v4.3/esp32c3/api-reference/peripherals/timer.html)

I believe the issue is within esp-idf actually, opened an [issue there](https://github.com/espressif/esp-idf/issues/15466).

**Comment by robert-hh** [comment]
As far as I understand the manual, the ESP32C3 has two general purpose timers, one in each of the two timer groups. There is another timer in each timer group, but that is a watchdog timer. 
Note that the code for machine_timer takes the number of timers from the ESP32 header files.

**Comment by projectgus** [comment]
> > @robert-hh @projectgus, I'm not so sure about this. I'm testing this now and the ESP32C3 loses access to two timers.
> > > [The two ESP32-C3 timer groups, with two timers in each, provide the total of four individual timers for use.](https://docs.espressif.com/projects/esp-idf/en/v4.3/esp32c3/api-reference/peripherals/timer.html)
> 
> I believe the issue is within esp-idf actually, opened an [issue there](https://github.com/espressif/esp-idf/issues/15466).

I agree with @robert-hh, I think the bug might be in that line of ESP-IDF documentation and not the ESP-IDF code.

Here's the ESP32-C3 reference manual:
image

**Comment by projectgus** [comment]
(That quoted line is also no longer present in the [latest version of the ESP-IDF docs](https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/peripherals/gptimer.html), the linked docs are for the v4.3 release which came out in 2021.)

**Comment by graeme44** [comment]
Many thanks. My mistake, I was looking at old docs – [confirmed here](https://github.com/espressif/esp-idf/issues/15466):

**Pull request merged**